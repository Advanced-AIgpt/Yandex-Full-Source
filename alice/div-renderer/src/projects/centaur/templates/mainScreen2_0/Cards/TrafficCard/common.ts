import { ITrafficForecast } from './types';

export function sortTimeOfDay(originalArray: Readonly<ITrafficForecast[]>): ITrafficForecast[] {
    const forecastsFirst = [];
    const forecastsLast = [];

    for (const forecastElement of originalArray) {
        if (forecastElement.time === null || typeof forecastElement.time === 'undefined') {
            continue;
        }

        if (forecastsFirst.length === 0) {
            forecastsFirst.push(forecastElement);
        } else if (forecastsLast.length > 0) {
            forecastsLast.push(forecastElement);
        } else if (forecastElement.time - (forecastsFirst[forecastsFirst.length - 1].time as number) === 1) {
            forecastsFirst.push(forecastElement);
        } else {
            forecastsLast.push(forecastElement);
        }
    }

    return [...forecastsLast, ...forecastsFirst];
}
