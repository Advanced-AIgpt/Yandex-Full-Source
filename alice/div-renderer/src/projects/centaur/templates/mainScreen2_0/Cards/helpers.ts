export function getCardMainScreenId({ rowIndex, colIndex }: { rowIndex?: number; colIndex?: number }) {
    return `main_screen_card_${rowIndex || 0}_${colIndex || 0}`;
}
