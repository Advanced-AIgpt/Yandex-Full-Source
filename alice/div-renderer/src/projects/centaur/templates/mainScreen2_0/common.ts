import { IDivAction } from 'divcard2';
import { EnumLayer } from '../../actions/client';
import { CrutchLoader } from '../../components/Loader/Loader';

interface IPutLoaderToActionOptions {
    action?: string | null,
    layer: EnumLayer,
    actionLogId?: string,
}

export const putLoaderToAction = ({
    action,
    layer,
    actionLogId = 'action_with_loader',
}: IPutLoaderToActionOptions): IDivAction[] | undefined => {
    if (!action) {
        return undefined;
    }
    return [
        CrutchLoader(layer),
        {
            log_id: actionLogId,
            url: action,
        },
    ];
};
