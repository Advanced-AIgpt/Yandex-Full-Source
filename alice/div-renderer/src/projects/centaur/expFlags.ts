import { createHasExperiment } from '../../common/helpers/expFlags';

export enum ExpFlags {
    musicPerformanceOpen = 'music_performance_open',
    newFact = 'divrender_new_fact',
    scenarioDataPrettyJSON = 'scenario_data_pretty_json',
    errorOnAssert = 'error_on_assert',
    showError = 'show_error',
    teasersDesignWithDoubleScreenNews = 'teasers_design_with_double_screen_news',
    teasersDesignWithDoubleScreenWeather = 'teasers_design_with_double_screen_weather',
    extendedNewsDesignWithDoubleScreen = 'extended_news_design_with_double_screen',
    extendedNewsDesignWithDoubleScreen2 = 'extended_news_design_with_double_screen2',
    searchRichCardSkeleton = 'search_rich_card_skeleton',
    templateMayNoExist = 'template_may_no_exist',
    scenarioWidgetMechanics = 'scenario_widget_mechanics',
    photoFrameDebugUpdateData = 'photo_frame_debug_update_data',
    div2CardAsString = 'div2_card_as_string',
    renderPlayerLikes = 'render_player_likes',
    telegramCallControlButtons = 'telegram_call_control_buttons',
    mainScreenTypedAction = 'centaur_typed_action',
    videoSearchShowAllResults = 'video_search_show_all_results',
}

export const hasExperiment = createHasExperiment<ExpFlags>();
