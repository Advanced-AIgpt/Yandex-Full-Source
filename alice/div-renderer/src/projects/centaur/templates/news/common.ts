import {
    ContainerBlock,
    DivBackground,
    FixedSize, IDivAction,
    ImageBackground,
    ImageBlock,
    MatchParentSize,
    SolidBackground,
    TextBlock,
    WrapContentSize,
} from 'divcard2';
import { compact } from 'lodash';
import { utcToZonedTime } from 'date-fns-tz';
import { differenceInCalendarDays, isSameDay } from 'date-fns';
import { Long } from 'protobufjs';
import { NAlice } from '../../../../protos';
import { formatDate } from '../../helpers/helpers';
import {
    colorDarkGrey1Opacity70,
    colorReadButton,
    colorWhite,
    colorWhiteOpacity60,
    offsetFromEdgeOfScreen,
} from '../../style/constants';
import { text40r, title32m, title60m } from '../../style/Text/Text';
import { getImageByTopic } from '../../helpers/newsHelper';
import { Avatar } from '../../../../common/helpers/avatar';

export const patchToOrig = (avatarsUrl: string) => avatarsUrl.replace(/\/[^/]+$/, '/orig');

export type NewsCardProps = {
    item: NAlice.NData.ITNewsItem;
    topic: string;
    bottomGap: boolean;
    tz: string;
};

export function newsCommonCard(
    { item: { Text, Image, Agency, Logo, PubDate, ExtendedNews }, topic, bottomGap, tz }: NewsCardProps,
    openExtendedNewsDivAction: IDivAction) {
    return new ContainerBlock({
        width: new MatchParentSize(),
        height: new MatchParentSize(),
        background: generateBackground(topic, Image),
        content_alignment_horizontal: 'left',
        content_alignment_vertical: 'bottom',
        paddings: {
            left: offsetFromEdgeOfScreen,
            bottom: bottomGap ? 104 : offsetFromEdgeOfScreen,
            right: offsetFromEdgeOfScreen,
        },
        items: [
            new ContainerBlock({
                width: new WrapContentSize(),
                height: new MatchParentSize({ weight: 1 }),
                content_alignment_vertical: 'bottom',
                orientation: 'horizontal',
                margins: {
                    bottom: 20,
                },
                items: newsSource(Agency, Logo, PubDate, tz),
            }),
            new ContainerBlock({
                width: new MatchParentSize(),
                height: new WrapContentSize(),
                content_alignment_vertical: 'bottom',
                orientation: 'horizontal',
                items: compact([
                    new TextBlock({
                        ...title60m,
                        width: new MatchParentSize({ weight: 1 }),
                        height: new WrapContentSize(),
                        max_lines: 5,
                        truncate: 'end',
                        text: Text ?? ' ',
                    }),
                    ExtendedNews?.length && new TextBlock({
                        ...text40r,
                        width: new WrapContentSize(),
                        height: new WrapContentSize(),
                        background: [new SolidBackground({
                            color: colorReadButton,
                        })],
                        border: {
                            corner_radius: 24,
                        },
                        paddings: {
                            top: 22,
                            right: 33,
                            bottom: 22,
                            left: 33,
                        },
                        margins: {
                            left: 48,
                        },
                        alignment_horizontal: 'right',
                        text: 'Читать',
                        action: openExtendedNewsDivAction,
                    }),
                ]),
            }),
        ],
    });
}

export const newsSource = (
    Agency: (string | null | undefined),
    Logo: (string | null | undefined),
    PubDate: (number | Long | null | undefined),
    tz: string,
) =>
    compact([
        Logo && new ContainerBlock({
            background: [
                new SolidBackground({
                    color: colorWhite,
                }),
            ],
            content_alignment_horizontal: 'center',
            content_alignment_vertical: 'center',
            border: {
                corner_radius: 12,
            },
            width: new FixedSize({ value: 40 }),
            height: new FixedSize({ value: 40 }),
            margins: {
                right: 20,
            },
            items: [
                new ImageBlock({
                    width: new FixedSize({ value: 28 }),
                    height: new FixedSize({ value: 28 }),
                    image_url: Logo ?? '',
                }),
            ],
        }),
        new TextBlock({
            ...title32m,
            text_color: colorWhiteOpacity60,
            width: new WrapContentSize(),
            height: new WrapContentSize(),
            text_alignment_vertical: 'center',
            text: (Agency ? Agency + ' • ' : '') + (Number(PubDate) ? formDate(Number(PubDate) * 1000 || 0, tz) : ''),
        }),
    ]);

export const generateBackground = (
    Topic: string | null | undefined,
    NewsImage: NAlice.NData.TNewsItem.ITImage | null | undefined,
    color = colorDarkGrey1Opacity70,
) => {
    const result: DivBackground[] = [];
    if (NewsImage && NewsImage.Src) {
        const avatar = Avatar.fromUrl(NewsImage.Src);

        result.push(
            new ImageBackground({
                image_url: avatar ? avatar.setTypeName('orig').toString() : NewsImage.Src,
                scale: 'fill',
                preload_required: 1,
            }),
        );
    } else {
        result.push(
            new ImageBackground({
                image_url: getImageByTopic(Topic),
                preload_required: 1,
            }),
        );
    }

    result.push(new SolidBackground({ color }));
    return result;
};

export const formDate = (date: number, tz: string) => {
    const d = utcToZonedTime(date, tz);
    const now = utcToZonedTime(Date.now(), tz);
    if (isSameDay(d, now)) {
        return formatDate(d, 'HH:mm');
    }
    if (differenceInCalendarDays(now, d) === 1) {
        return `вчера в ${formatDate(d, 'HH:mm')}`;
    }
    return formatDate(d, 'd MMMM в HH:mm');
};
