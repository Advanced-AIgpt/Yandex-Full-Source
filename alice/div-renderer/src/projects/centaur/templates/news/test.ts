// import teasersRender from './teasers';
import scenarioRender from './scenario';
import { AnonymizeDataForSnapshot } from '../../../../common/helpers/dev';
import { MMRequest } from '../../../../common/helpers/MMRequest';
import { createRequestState } from '../../../../registries/common';
import { ExpFlags } from '../../expFlags';

jest.mock('../../../../common/logger');

// TODO: вернуть тест когда с сервера начнет приходить server_time
// describe('News teaser', () => {
//     it('matches snapshot', () => {
//         expect(AnonymizeDataForSnapshot(teasersRender({
//             Topic: 'sport',
//             NewsItem: {
//                 Agency: 'agency',
//                 Url: 'https://yandex.ru/news/1',
//                 Image: {
//                     Src: 'https://avatars.mds.yandex.net/get-bass/image/60x60',
//                 },
//                 Text: 'text',
//                 PubDate: 1634130001,
//                 Logo: 'https://avatars.mds.yandex.net/get-ynews-logo/26056/1071-1582024400668-square/orig',
//             },
//             Tz: 'Europe/Moscow',
//         }))).toMatchSnapshot();
//     });
// });

describe('News scenario', () => {
    it('matches snapshot', () => {
        expect(AnonymizeDataForSnapshot(scenarioRender({
            CurrentNewsItem: 0,
            NewsItems: [{
                Agency: 'agency',
                Image: {
                    Src: 'https://avatars.mds.yandex.net/get-bass/image/60x60',
                },
                Logo: 'imager',
                PubDate: 1000000,
                Messages: 'messages',
                Text: 'text',
                TurboIconUrl: 'https://avatars.mds.yandex.net/get-bass/turbo/32x32',
                Url: 'https://yandex.ru/news/1',
            }],
            Topic: 'world',
        }, new MMRequest({}, {}, {}), createRequestState()))).toMatchSnapshot();
    });
});

describe('News scenario with extended news', () => {
    const mmRequest = new MMRequest({}, {}, {
        [ExpFlags.extendedNewsDesignWithDoubleScreen as string]: '',
    });

    it('matches snapshot', () => {
        expect(AnonymizeDataForSnapshot(scenarioRender({
            CurrentNewsItem: 0,
            NewsItems: [{
                Agency: 'agency',
                Image: {
                    Src: 'https://avatars.mds.yandex.net/get-bass/image/60x60',
                },
                Logo: 'imager',
                PubDate: 1000000,
                Messages: 'messages',
                Text: 'text',
                TurboIconUrl: 'https://avatars.mds.yandex.net/get-bass/turbo/32x32',
                Url: 'https://yandex.ru/news/1',
            }],
            Topic: 'world',
        }, mmRequest, createRequestState(mmRequest)))).toMatchSnapshot();
    });
});
