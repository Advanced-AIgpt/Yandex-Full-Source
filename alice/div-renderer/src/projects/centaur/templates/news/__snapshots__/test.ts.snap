// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`News scenario matches snapshot 1`] = `
Object {
  "card": Object {
    "log_id": "news_scenario",
    "states": Array [
      Object {
        "div": Object {
          "default_state_id": "pager",
          "div_id": "news",
          "height": Object {
            "type": "match_parent",
          },
          "states": Array [
            Object {
              "animation_in": Object {
                "duration": 200,
                "end_value": 1,
                "interpolator": "ease_in_out",
                "name": "scale",
                "start_value": 0.9,
              },
              "animation_out": Object {
                "duration": 200,
                "end_value": 0.95,
                "interpolator": "ease_out",
                "name": "scale",
                "start_value": 1,
              },
              "div": Object {
                "height": Object {
                  "type": "match_parent",
                },
                "items": Array [
                  Object {
                    "default_item": 0,
                    "height": Object {
                      "type": "match_parent",
                    },
                    "id": "pager",
                    "items": Array [
                      Object {
                        "background": Array [
                          Object {
                            "image_url": "https://avatars.mds.yandex.net/get-bass/image/60x60",
                            "preload_required": 1,
                            "scale": "fill",
                            "type": "image",
                          },
                          Object {
                            "color": "#b3121316",
                            "type": "solid",
                          },
                        ],
                        "content_alignment_horizontal": "left",
                        "content_alignment_vertical": "bottom",
                        "height": Object {
                          "type": "match_parent",
                        },
                        "items": Array [
                          Object {
                            "content_alignment_vertical": "bottom",
                            "height": Object {
                              "type": "match_parent",
                              "weight": 1,
                            },
                            "items": Array [
                              Object {
                                "background": Array [
                                  Object {
                                    "color": "#fff",
                                    "type": "solid",
                                  },
                                ],
                                "border": Object {
                                  "corner_radius": 12,
                                },
                                "content_alignment_horizontal": "center",
                                "content_alignment_vertical": "center",
                                "height": Object {
                                  "type": "fixed",
                                  "value": 40,
                                },
                                "items": Array [
                                  Object {
                                    "height": Object {
                                      "type": "fixed",
                                      "value": 28,
                                    },
                                    "image_url": "imager",
                                    "type": "image",
                                    "width": Object {
                                      "type": "fixed",
                                      "value": 28,
                                    },
                                  },
                                ],
                                "margins": Object {
                                  "right": 20,
                                },
                                "type": "container",
                                "width": Object {
                                  "type": "fixed",
                                  "value": 40,
                                },
                              },
                              Object {
                                "font_size": 32,
                                "font_weight": "medium",
                                "height": Object {
                                  "type": "wrap_content",
                                },
                                "line_height": 40,
                                "text": "agency • 12 января в 13:46",
                                "text_alignment_vertical": "center",
                                "text_color": "#9fff",
                                "type": "text",
                                "width": Object {
                                  "type": "wrap_content",
                                },
                              },
                            ],
                            "margins": Object {
                              "bottom": 20,
                            },
                            "orientation": "horizontal",
                            "type": "container",
                            "width": Object {
                              "type": "wrap_content",
                            },
                          },
                          Object {
                            "content_alignment_vertical": "bottom",
                            "height": Object {
                              "type": "wrap_content",
                            },
                            "items": Array [
                              Object {
                                "font_size": 60,
                                "font_weight": "medium",
                                "height": Object {
                                  "type": "wrap_content",
                                },
                                "line_height": 72,
                                "max_lines": 5,
                                "text": "text",
                                "text_color": "#fff",
                                "truncate": "end",
                                "type": "text",
                                "width": Object {
                                  "type": "match_parent",
                                  "weight": 1,
                                },
                              },
                            ],
                            "orientation": "horizontal",
                            "type": "container",
                            "width": Object {
                              "type": "match_parent",
                            },
                          },
                        ],
                        "paddings": Object {
                          "bottom": 104,
                          "left": 48,
                          "right": 48,
                        },
                        "type": "container",
                        "width": Object {
                          "type": "match_parent",
                        },
                      },
                    ],
                    "layout_mode": Object {
                      "page_width": Object {
                        "type": "percentage",
                        "value": 100,
                      },
                      "type": "percentage",
                    },
                    "type": "pager",
                    "width": Object {
                      "type": "match_parent",
                    },
                  },
                  Object {
                    "content_alignment_horizontal": "center",
                    "content_alignment_vertical": "bottom",
                    "height": Object {
                      "type": "match_parent",
                    },
                    "items": Array [
                      Object {
                        "active_item_color": "#fff",
                        "height": Object {
                          "type": "wrap_content",
                        },
                        "inactive_item_color": "#1affffff",
                        "pager_id": "pager",
                        "shape": Object {
                          "corner_radius": Object {
                            "type": "fixed",
                            "value": 20,
                          },
                          "item_height": Object {
                            "type": "fixed",
                            "value": 8,
                          },
                          "item_width": Object {
                            "type": "fixed",
                            "value": 90,
                          },
                          "type": "rounded_rectangle",
                        },
                        "space_between_centers": Object {
                          "type": "fixed",
                          "value": 120,
                        },
                        "type": "indicator",
                        "width": Object {
                          "type": "match_parent",
                        },
                      },
                    ],
                    "paddings": Object {
                      "bottom": 48,
                    },
                    "type": "container",
                    "width": Object {
                      "type": "match_parent",
                    },
                  },
                  Object {
                    "actions": Array [
                      Object {
                        "log_id": "close_fullscreen",
                        "url": "centaur://local_command?local_commands=[%7B%22command%22%3A%22close_layer%22%2C%22layer%22%3A%22DIALOG%22%2C%22stop_conversation%22%3Atrue%7D]",
                      },
                    ],
                    "alignment_horizontal": "right",
                    "background": Array [
                      Object {
                        "color": "#1affffff",
                        "type": "solid",
                      },
                    ],
                    "border": Object {
                      "corner_radius": 20,
                    },
                    "height": Object {
                      "type": "fixed",
                      "value": 88,
                    },
                    "image_url": "https://yastatic.net/s3/dialogs/smart_displays/icons/close_wt_border.png",
                    "margins": Object {
                      "right": 48,
                      "top": 48,
                    },
                    "paddings": Object {
                      "bottom": 10,
                      "left": 10,
                      "right": 10,
                      "top": 10,
                    },
                    "preload_required": 1,
                    "scale": "fill",
                    "type": "image",
                    "width": Object {
                      "type": "fixed",
                      "value": 88,
                    },
                  },
                ],
                "orientation": "overlap",
                "type": "container",
                "width": Object {
                  "type": "match_parent",
                },
              },
              "state_id": "pager",
            },
            Object {
              "animation_in": Object {
                "duration": 350,
                "end_value": 0,
                "interpolator": "ease_out",
                "name": "translate",
                "start_value": 1,
              },
              "animation_out": Object {
                "duration": 350,
                "end_value": 1,
                "interpolator": "ease_in",
                "name": "translate",
                "start_value": 0,
              },
              "div": Object {
                "background": Array [
                  Object {
                    "color": "#121316",
                    "type": "solid",
                  },
                ],
                "height": Object {
                  "type": "match_parent",
                },
                "items": Array [
                  Object {
                    "actions": Array [
                      Object {
                        "log_id": "close_fullscreen",
                        "url": "div-action://set_state?state_id=0/news/pager&temporary=true",
                      },
                    ],
                    "alignment_horizontal": "right",
                    "background": Array [
                      Object {
                        "color": "#1affffff",
                        "type": "solid",
                      },
                    ],
                    "border": Object {
                      "corner_radius": 20,
                    },
                    "height": Object {
                      "type": "fixed",
                      "value": 88,
                    },
                    "image_url": "https://yastatic.net/s3/dialogs/smart_displays/icons/close_wt_border.png",
                    "margins": Object {
                      "top": 48,
                    },
                    "paddings": Object {
                      "bottom": 10,
                      "left": 10,
                      "right": 10,
                      "top": 10,
                    },
                    "preload_required": 1,
                    "scale": "fill",
                    "type": "image",
                    "width": Object {
                      "type": "fixed",
                      "value": 88,
                    },
                  },
                  Object {
                    "height": Object {
                      "type": "wrap_content",
                    },
                    "items": Array [
                      Object {
                        "height": Object {
                          "type": "wrap_content",
                        },
                        "items": Array [
                          Object {
                            "background": Array [
                              Object {
                                "color": "#fff",
                                "type": "solid",
                              },
                            ],
                            "border": Object {
                              "corner_radius": 12,
                            },
                            "content_alignment_horizontal": "center",
                            "content_alignment_vertical": "center",
                            "height": Object {
                              "type": "fixed",
                              "value": 40,
                            },
                            "items": Array [
                              Object {
                                "height": Object {
                                  "type": "fixed",
                                  "value": 28,
                                },
                                "image_url": "imager",
                                "type": "image",
                                "width": Object {
                                  "type": "fixed",
                                  "value": 28,
                                },
                              },
                            ],
                            "margins": Object {
                              "right": 20,
                            },
                            "type": "container",
                            "width": Object {
                              "type": "fixed",
                              "value": 40,
                            },
                          },
                          Object {
                            "font_size": 32,
                            "font_weight": "medium",
                            "height": Object {
                              "type": "wrap_content",
                            },
                            "line_height": 40,
                            "text": "agency • 12 января в 13:46",
                            "text_alignment_vertical": "center",
                            "text_color": "#9fff",
                            "type": "text",
                            "width": Object {
                              "type": "wrap_content",
                            },
                          },
                        ],
                        "margins": Object {
                          "bottom": 24,
                          "top": 48,
                        },
                        "orientation": "horizontal",
                        "type": "container",
                        "width": Object {
                          "type": "wrap_content",
                        },
                      },
                      Object {
                        "font_size": 60,
                        "font_weight": "medium",
                        "height": Object {
                          "type": "wrap_content",
                        },
                        "line_height": 72,
                        "margins": Object {
                          "bottom": 32,
                        },
                        "text": "text",
                        "text_color": "#fff",
                        "truncate": "end",
                        "type": "text",
                        "width": Object {
                          "type": "wrap_content",
                        },
                      },
                    ],
                    "margins": Object {
                      "right": 204,
                    },
                    "orientation": "vertical",
                    "type": "gallery",
                    "width": Object {
                      "type": "wrap_content",
                    },
                  },
                ],
                "orientation": "overlap",
                "paddings": Object {
                  "left": 48,
                  "right": 48,
                },
                "type": "container",
                "width": Object {
                  "type": "match_parent",
                },
              },
              "state_id": "extended_0",
            },
          ],
          "type": "state",
          "width": Object {
            "type": "match_parent",
          },
        },
        "state_id": 0,
      },
    ],
    "variable_triggers": Array [],
    "variables": Array [],
  },
  "templates": Object {},
}
`;

exports[`News scenario with extended news matches snapshot 1`] = `
Object {
  "card": Object {
    "log_id": "news_scenario",
    "states": Array [
      Object {
        "div": Object {
          "default_state_id": "pager",
          "div_id": "news",
          "height": Object {
            "type": "match_parent",
          },
          "states": Array [
            Object {
              "animation_in": Object {
                "duration": 200,
                "end_value": 1,
                "interpolator": "ease_in_out",
                "name": "scale",
                "start_value": 0.9,
              },
              "animation_out": Object {
                "duration": 200,
                "end_value": 0.95,
                "interpolator": "ease_out",
                "name": "scale",
                "start_value": 1,
              },
              "div": Object {
                "height": Object {
                  "type": "match_parent",
                },
                "items": Array [
                  Object {
                    "default_item": 0,
                    "height": Object {
                      "type": "match_parent",
                    },
                    "id": "pager",
                    "items": Array [
                      Object {
                        "background": Array [
                          Object {
                            "image_url": "https://avatars.mds.yandex.net/get-bass/image/60x60",
                            "preload_required": 1,
                            "scale": "fill",
                            "type": "image",
                          },
                          Object {
                            "color": "#b3121316",
                            "type": "solid",
                          },
                        ],
                        "content_alignment_horizontal": "left",
                        "content_alignment_vertical": "bottom",
                        "height": Object {
                          "type": "match_parent",
                        },
                        "items": Array [
                          Object {
                            "content_alignment_vertical": "bottom",
                            "height": Object {
                              "type": "match_parent",
                              "weight": 1,
                            },
                            "items": Array [
                              Object {
                                "background": Array [
                                  Object {
                                    "color": "#fff",
                                    "type": "solid",
                                  },
                                ],
                                "border": Object {
                                  "corner_radius": 12,
                                },
                                "content_alignment_horizontal": "center",
                                "content_alignment_vertical": "center",
                                "height": Object {
                                  "type": "fixed",
                                  "value": 40,
                                },
                                "items": Array [
                                  Object {
                                    "height": Object {
                                      "type": "fixed",
                                      "value": 28,
                                    },
                                    "image_url": "imager",
                                    "type": "image",
                                    "width": Object {
                                      "type": "fixed",
                                      "value": 28,
                                    },
                                  },
                                ],
                                "margins": Object {
                                  "right": 20,
                                },
                                "type": "container",
                                "width": Object {
                                  "type": "fixed",
                                  "value": 40,
                                },
                              },
                              Object {
                                "font_size": 32,
                                "font_weight": "medium",
                                "height": Object {
                                  "type": "wrap_content",
                                },
                                "line_height": 40,
                                "text": "agency • 12 января в 13:46",
                                "text_alignment_vertical": "center",
                                "text_color": "#9fff",
                                "type": "text",
                                "width": Object {
                                  "type": "wrap_content",
                                },
                              },
                            ],
                            "margins": Object {
                              "bottom": 20,
                            },
                            "orientation": "horizontal",
                            "type": "container",
                            "width": Object {
                              "type": "wrap_content",
                            },
                          },
                          Object {
                            "content_alignment_vertical": "bottom",
                            "height": Object {
                              "type": "wrap_content",
                            },
                            "items": Array [
                              Object {
                                "font_size": 60,
                                "font_weight": "medium",
                                "height": Object {
                                  "type": "wrap_content",
                                },
                                "line_height": 72,
                                "max_lines": 5,
                                "text": "text",
                                "text_color": "#fff",
                                "truncate": "end",
                                "type": "text",
                                "width": Object {
                                  "type": "match_parent",
                                  "weight": 1,
                                },
                              },
                            ],
                            "orientation": "horizontal",
                            "type": "container",
                            "width": Object {
                              "type": "match_parent",
                            },
                          },
                        ],
                        "paddings": Object {
                          "bottom": 104,
                          "left": 48,
                          "right": 48,
                        },
                        "type": "container",
                        "width": Object {
                          "type": "match_parent",
                        },
                      },
                    ],
                    "layout_mode": Object {
                      "page_width": Object {
                        "type": "percentage",
                        "value": 100,
                      },
                      "type": "percentage",
                    },
                    "type": "pager",
                    "width": Object {
                      "type": "match_parent",
                    },
                  },
                  Object {
                    "content_alignment_horizontal": "center",
                    "content_alignment_vertical": "bottom",
                    "height": Object {
                      "type": "match_parent",
                    },
                    "items": Array [
                      Object {
                        "active_item_color": "#fff",
                        "height": Object {
                          "type": "wrap_content",
                        },
                        "inactive_item_color": "#1affffff",
                        "pager_id": "pager",
                        "shape": Object {
                          "corner_radius": Object {
                            "type": "fixed",
                            "value": 20,
                          },
                          "item_height": Object {
                            "type": "fixed",
                            "value": 8,
                          },
                          "item_width": Object {
                            "type": "fixed",
                            "value": 90,
                          },
                          "type": "rounded_rectangle",
                        },
                        "space_between_centers": Object {
                          "type": "fixed",
                          "value": 120,
                        },
                        "type": "indicator",
                        "width": Object {
                          "type": "match_parent",
                        },
                      },
                    ],
                    "paddings": Object {
                      "bottom": 48,
                    },
                    "type": "container",
                    "width": Object {
                      "type": "match_parent",
                    },
                  },
                  Object {
                    "actions": Array [
                      Object {
                        "log_id": "close_fullscreen",
                        "url": "centaur://local_command?local_commands=[%7B%22command%22%3A%22close_layer%22%2C%22layer%22%3A%22DIALOG%22%2C%22stop_conversation%22%3Atrue%7D]",
                      },
                    ],
                    "alignment_horizontal": "right",
                    "background": Array [
                      Object {
                        "color": "#1affffff",
                        "type": "solid",
                      },
                    ],
                    "border": Object {
                      "corner_radius": 20,
                    },
                    "height": Object {
                      "type": "fixed",
                      "value": 88,
                    },
                    "image_url": "https://yastatic.net/s3/dialogs/smart_displays/icons/close_wt_border.png",
                    "margins": Object {
                      "right": 48,
                      "top": 48,
                    },
                    "paddings": Object {
                      "bottom": 10,
                      "left": 10,
                      "right": 10,
                      "top": 10,
                    },
                    "preload_required": 1,
                    "scale": "fill",
                    "type": "image",
                    "width": Object {
                      "type": "fixed",
                      "value": 88,
                    },
                  },
                ],
                "orientation": "overlap",
                "type": "container",
                "width": Object {
                  "type": "match_parent",
                },
              },
              "state_id": "pager",
            },
            Object {
              "animation_in": Object {
                "duration": 350,
                "end_value": 0,
                "interpolator": "ease_out",
                "name": "translate",
                "start_value": 1,
              },
              "animation_out": Object {
                "duration": 350,
                "end_value": 1,
                "interpolator": "ease_in",
                "name": "translate",
                "start_value": 0,
              },
              "div": Object {
                "background": Array [
                  Object {
                    "color": "#121316",
                    "type": "solid",
                  },
                ],
                "height": Object {
                  "type": "match_parent",
                },
                "items": Array [
                  Object {
                    "actions": Array [
                      Object {
                        "log_id": "close_fullscreen",
                        "url": "div-action://set_state?state_id=0/news/pager&temporary=true",
                      },
                    ],
                    "alignment_horizontal": "right",
                    "background": Array [
                      Object {
                        "color": "#1affffff",
                        "type": "solid",
                      },
                    ],
                    "border": Object {
                      "corner_radius": 20,
                    },
                    "height": Object {
                      "type": "fixed",
                      "value": 88,
                    },
                    "image_url": "https://yastatic.net/s3/dialogs/smart_displays/icons/close_wt_border.png",
                    "margins": Object {
                      "top": 48,
                    },
                    "paddings": Object {
                      "bottom": 10,
                      "left": 10,
                      "right": 10,
                      "top": 10,
                    },
                    "preload_required": 1,
                    "scale": "fill",
                    "type": "image",
                    "width": Object {
                      "type": "fixed",
                      "value": 88,
                    },
                  },
                  Object {
                    "height": Object {
                      "type": "wrap_content",
                    },
                    "items": Array [
                      Object {
                        "height": Object {
                          "type": "wrap_content",
                        },
                        "items": Array [
                          Object {
                            "background": Array [
                              Object {
                                "color": "#fff",
                                "type": "solid",
                              },
                            ],
                            "border": Object {
                              "corner_radius": 12,
                            },
                            "content_alignment_horizontal": "center",
                            "content_alignment_vertical": "center",
                            "height": Object {
                              "type": "fixed",
                              "value": 40,
                            },
                            "items": Array [
                              Object {
                                "height": Object {
                                  "type": "fixed",
                                  "value": 28,
                                },
                                "image_url": "imager",
                                "type": "image",
                                "width": Object {
                                  "type": "fixed",
                                  "value": 28,
                                },
                              },
                            ],
                            "margins": Object {
                              "right": 20,
                            },
                            "type": "container",
                            "width": Object {
                              "type": "fixed",
                              "value": 40,
                            },
                          },
                          Object {
                            "font_size": 32,
                            "font_weight": "medium",
                            "height": Object {
                              "type": "wrap_content",
                            },
                            "line_height": 40,
                            "text": "agency • 12 января в 13:46",
                            "text_alignment_vertical": "center",
                            "text_color": "#9fff",
                            "type": "text",
                            "width": Object {
                              "type": "wrap_content",
                            },
                          },
                        ],
                        "margins": Object {
                          "bottom": 24,
                          "top": 48,
                        },
                        "orientation": "horizontal",
                        "type": "container",
                        "width": Object {
                          "type": "wrap_content",
                        },
                      },
                      Object {
                        "font_size": 60,
                        "font_weight": "medium",
                        "height": Object {
                          "type": "wrap_content",
                        },
                        "line_height": 72,
                        "margins": Object {
                          "bottom": 32,
                        },
                        "text": "text",
                        "text_color": "#fff",
                        "truncate": "end",
                        "type": "text",
                        "width": Object {
                          "type": "wrap_content",
                        },
                      },
                    ],
                    "margins": Object {
                      "right": 204,
                    },
                    "orientation": "vertical",
                    "type": "gallery",
                    "width": Object {
                      "type": "wrap_content",
                    },
                  },
                ],
                "orientation": "overlap",
                "paddings": Object {
                  "left": 48,
                  "right": 48,
                },
                "type": "container",
                "width": Object {
                  "type": "match_parent",
                },
              },
              "state_id": "extended_0",
            },
          ],
          "type": "state",
          "width": Object {
            "type": "match_parent",
          },
        },
        "state_id": 0,
      },
    ],
    "variable_triggers": Array [],
    "variables": Array [],
  },
  "templates": Object {},
}
`;
